src/v/storage/disk_log_impl.cc

```diff
@@ -1117,91 +1118,79 @@ ss::future<> disk_log_impl::truncate(truncate_config cfg) {
 
 ss::future<> disk_log_impl::do_truncate(truncate_config cfg) {
     auto stats = offsets();
-    if (cfg.base_offset > stats.dirty_offset) {
-        return ss::make_ready_future<>();
-    }
-    if (_segs.empty()) {
-        return ss::make_ready_future<>();
+
+    if (cfg.base_offset > stats.dirty_offset || _segs.empty()) {
+        co_return;
     }
+
     cfg.base_offset = std::max(cfg.base_offset, _start_offset);
     // Note different from the stats variable above because
     // we want to delete even empty segments.
     if (
       cfg.base_offset
       < std::max(_segs.back()->offsets().base_offset, _start_offset)) {
-        return remove_full_segments(cfg.base_offset).then([this, cfg] {
-            // recurse
-            return do_truncate(cfg);
-        });
+        co_await remove_full_segments(cfg.base_offset);
+        // recurse
+        co_return co_await do_truncate(cfg);
     }
-    auto& last = *_segs.back();
-    if (cfg.base_offset > last.offsets().dirty_offset) {
-        return ss::make_ready_future<>();
+    auto last = _segs.back();
+    if (cfg.base_offset > last->offsets().dirty_offset) {
+        co_return;
     }
-    auto pidx = last.index().find_nearest(cfg.base_offset);
-    model::offset start = last.index().base_offset();
+
+    co_await last->flush();
```

Comment:
in the non-coroutine version this flush comes _after_ `find_nearest` and `base_offset()` (the next two lines). is that intentional?

Reply:
yes, this is intentional change, we want to query offsets after a flush